src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs

```diff
@@ -659,6 +647,18 @@ public static void Run(ServiceBase[] services)
             }
             finally
             {
+                // Free the pointer to the name of the service on the unmanaged heap.
+                // We don't need to free the last element in the unmanaged array since the last element have null values (denotes the end of the table).
+                for (int i = 0; i < services.Length; i++)
+                {
+                    SERVICE_TABLE_ENTRY entry = Marshal.PtrToStructure<SERVICE_TABLE_ENTRY>(entriesPointer + sizeOfSERVICE_TABLE_ENTRY * i);
```

Comment:
` Marshal.PtrToStructure` is also going to marshal the delegate inside SERVICE_TABLE_ENTRY (non-trivial operation). It is very wasteful to marshal the delegate just to free a bit of memory.

Reply:
I agree. I think we could use a Span to read directly from it